File: shifai-ios/ShifAI/Data/FeatureFlags.swift
import Foundation

/// Feature Flags — remote config for gradual feature rollout.
/// Defaults to local values, overridden by Supabase app_config table.
final class FeatureFlags {

    static let shared = FeatureFlags()

    private var remoteFlags: [String: Bool] = [:]
    private let defaults: [String: Bool] = [
        "ml_predictions": false,
        "share_links": true,
        "cycle_insights": true,
        "body_map_v2": false,
        "pdf_export": true,
        "widget_predictions": false,
        "biometric_lock": true,
        "analytics_v2": false,
        "background_sync": true,
        "csv_export": true,
    ]

    private init() {}

    // MARK: - Access

    func isEnabled(_ flag: String) -> Bool {
        remoteFlags[flag] ?? defaults[flag] ?? false
    }

    var mlPredictions: Bool { isEnabled("ml_predictions") }
    var shareLinks: Bool { isEnabled("share_links") }
    var cycleInsights: Bool { isEnabled("cycle_insights") }
    var bodyMapV2: Bool { isEnabled("body_map_v2") }
    var pdfExport: Bool { isEnabled("pdf_export") }
    var widgetPredictions: Bool { isEnabled("widget_predictions") }
    var biometricLock: Bool { isEnabled("biometric_lock") }
    var backgroundSync: Bool { isEnabled("background_sync") }
    var csvExport: Bool { isEnabled("csv_export") }

    // MARK: - Remote Update

    func update(from remote: [String: Bool]) {
        remoteFlags = remote
    }

    func reset() {
        remoteFlags.removeAll()
    }
}

File: shifai-android/app/src/main/java/com/shifai/data/config/FeatureFlags.kt
package com.shifai.data.config

/**
 * Feature Flags — remote config for gradual feature rollout.
 * Defaults to local values, overridden by Supabase app_config table.
 * Mirrors iOS FeatureFlags.swift.
 */
object FeatureFlags {

    private var remoteFlags = mutableMapOf<String, Boolean>()
    private val defaults = mapOf(
        "ml_predictions" to false,
        "share_links" to true,
        "cycle_insights" to true,
        "body_map_v2" to false,
        "pdf_export" to true,
        "widget_predictions" to false,
        "biometric_lock" to true,
        "analytics_v2" to false,
        "background_sync" to true,
        "csv_export" to true,
    )

    // ─── Access ───

    fun isEnabled(flag: String): Boolean =
        remoteFlags[flag] ?: defaults[flag] ?: false

    val mlPredictions: Boolean get() = isEnabled("ml_predictions")
    val shareLinks: Boolean get() = isEnabled("share_links")
    val cycleInsights: Boolean get() = isEnabled("cycle_insights")
    val bodyMapV2: Boolean get() = isEnabled("body_map_v2")
    val pdfExport: Boolean get() = isEnabled("pdf_export")
    val widgetPredictions: Boolean get() = isEnabled("widget_predictions")
    val biometricLock: Boolean get() = isEnabled("biometric_lock")
    val backgroundSync: Boolean get() = isEnabled("background_sync")
    val csvExport: Boolean get() = isEnabled("csv_export")

    // ─── Remote Update ───

    fun update(remote: Map<String, Boolean>) {
        remoteFlags.clear()
        remoteFlags.putAll(remote)
    }

    fun reset() {
        remoteFlags.clear()
    }
}
